filebeat/inputsource/common/dgram/server.go

```diff
@@ -0,0 +1,194 @@
+	path string,
+	connect HandlerFactory,
+	listenerFactory ListenerFactory,
+	config *ListenerConfig,
+) *Listener {
+	return &Listener{
+		log:      logp.NewLogger(f.String()),
+		config:   config,
+		listener: listenerFactory,
+		connect:  connect,
+		tg:       unison.TaskGroup{},
+	}
+}
+
+// DatagramReaderFactory allows creation of a handler that has splitting capabilities.
+func DatagramReaderFactory(
+	family inputsource.Family,
+	logger *logp.Logger,
+	callback inputsource.NetworkFunc,
+) HandlerFactory {
+	return func(config ListenerConfig) ConnectionHandler {
+		return ConnectionHandler(func(ctx context.Context, conn ReadFromConnection) error {
+			//metadata := metadataCallback(conn)
+			for ctx.Err() == nil {
+
+				buffer := make([]byte, config.MaxMessageSize)
+				conn.SetDeadline(time.Now().Add(config.Timeout))
+
+				// If you are using Windows and you are using a fixed buffer and you get a datagram which
+				// is bigger than the specified size of the buffer, it will return an `err` and the buffer will
+				// contains a subset of the data.
+				//
+				// On Unix based system, the buffer will be truncated but no error will be returned.
+				length, addr, err := conn.ReadFrom(buffer)
+				if err != nil {
+					// don't log any deadline events.
+					e, ok := err.(net.Error)
+					if ok && e.Timeout() {
+						continue
+					}
+
+					// Closed network error string will never change in Go 1.X
+					// https://github.com/golang/go/issues/4373
+					opErr, ok := err.(*net.OpError)
+					if ok && strings.Contains(opErr.Err.Error(), "use of closed network connection") {
+						logger.Info("Connection has been closed")
+						return nil
+					}
+
+					logger.Errorf("Error reading from the socket %s", err)
+
+					// On Windows send the current buffer and mark it as truncated.
+					// The buffer will have content but length will return 0, addr will be nil.
+					if isLargerThanBuffer(err) {
+						callback(buffer, inputsource.NetworkMetadata{RemoteAddr: addr, Truncated: true})
+						continue
+					}
+				}
+
+				if length > 0 {
+					callback(buffer[:length], inputsource.NetworkMetadata{RemoteAddr: addr})
+				}
+			}
+			return nil
+		})
+	}
+}
+
+func isLargerThanBuffer(err error) bool {
+	if runtime.GOOS != "windows" {
+		return false
+	}
+	return strings.Contains(err.Error(), windowErrBuffer)
+}
+
+func (l *Listener) Run(ctx context.Context) error {
+	conn, err := l.listener()
+	if err != nil {
+		return err
+	}
+	connCtx, connCancel := ctxtool.WithFunc(ctxtool.FromCanceller(ctx), func() {
+		conn.Close()
+	})
+	defer connCancel()
+
+	handler := l.connect(*l.config)
+	for connCtx.Err() == nil {
+		err := handler(connCtx, conn)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (l *Listener) Start() error {
+	conn, err := l.listener()
+	if err != nil {
+		return err
+	}
+
+	l.tg.Go(func(ctx unison.Canceler) error {
+		connCtx, connCancel := ctxtool.WithFunc(ctxtool.FromCanceller(ctx), func() {
```

Comment:
This looks similar to run. Do we want 'Start' to call Run or have a common `func (l *Listener) readLoop(...)`?

What about error handling in `Start`. If listener fails would it make sense to retry similar to `Run` and only fail initially if the error is really non-recoverable?

Reply:
> What about error handling in Start. If listener fails would it make sense to retry similar to Run and only fail initially if the error is really non-recoverable?

To me `Start` sounds like a function you would expect it try to start something and return if it cannot. `Run` is more like constantly retrying to do something.

Reply:
This is why "non-recoverable". If the error can be recovered from I would expect the internal runLoop to continue trying to connect and collect data if possible. Otherwise I will be forced to monitor the health of each input (not really supported yet) and restart filebeat until it eventually went healthy.

If some component implements `Start` and `Run`, I would not expect them to behave differently depending on which method I use (beside Run should block and Start return immediately).